File: modules/audit-log/backend/src/services/audit.service.ts
import { PrismaClient } from '@prisma/client';

// =============================================================================
// Types
// =============================================================================

export type AuditLevel = 'info' | 'warning' | 'error' | 'security';

export type AuditCategory =
  | 'auth'
  | 'user'
  | 'admin'
  | 'payment'
  | 'system'
  | 'api'
  | 'security'
  | string;

export interface AuditLogEntry {
  level?: AuditLevel;
  action: string;
  category?: AuditCategory;
  userId?: string;
  userEmail?: string;
  targetId?: string;
  targetType?: string;
  method?: string;
  path?: string;
  statusCode?: number;
  ipAddress?: string;
  userAgent?: string;
  duration?: number;
  metadata?: Record<string, unknown>;
  error?: string;
}

export interface AuditLogRecord extends AuditLogEntry {
  id: string;
  timestamp: Date;
}

export interface AuditQueryOptions {
  startDate?: Date;
  endDate?: Date;
  userId?: string;
  level?: AuditLevel;
  levels?: AuditLevel[];
  action?: string;
  actions?: string[];
  category?: AuditCategory;
  categories?: AuditCategory[];
  search?: string;
  page?: number;
  limit?: number;
  sortBy?: 'timestamp' | 'level' | 'action';
  sortOrder?: 'asc' | 'desc';
}

export interface AuditQueryResult {
  logs: AuditLogRecord[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export interface AuditStats {
  total: number;
  byLevel: Record<AuditLevel, number>;
  byCategory: Record<string, number>;
  recentActivity: Array<{
    date: string;
    count: number;
  }>;
}

export interface AuditConfig {
  enabled: boolean;
  retentionDays: number;
  levels: AuditLevel[];
  excludePaths: RegExp[];
  includeBody: boolean;
  maskFields: string[];
}

// =============================================================================
// Audit Service
// =============================================================================

export class AuditService {
  private config: AuditConfig;
  private queue: AuditLogEntry[] = [];
  private flushInterval: ReturnType<typeof setInterval> | null = null;
  private readonly BATCH_SIZE = 100;
  private readonly FLUSH_INTERVAL_MS = 5000;
  private prisma: PrismaClient;

  constructor(config?: Partial<AuditConfig>, prisma?: PrismaClient) {
    this.prisma = prisma || new PrismaClient();
    this.config = {
      enabled: config?.enabled ?? process.env.AUDIT_LOG_ENABLED !== 'false',
      retentionDays: config?.retentionDays ?? parseInt(process.env.AUDIT_LOG_RETENTION_DAYS || '365'),
      levels: config?.levels ?? this.parseLevels(process.env.AUDIT_LOG_LEVELS),
      excludePaths: config?.excludePaths ?? this.parsePaths(process.env.AUDIT_LOG_EXCLUDE_PATHS),
      includeBody: config?.includeBody ?? process.env.AUDIT_LOG_INCLUDE_BODY === 'true',
      maskFields: config?.maskFields ?? this.parseFields(process.env.AUDIT_LOG_MASK_FIELDS),
    };

    if (this.config.enabled) {
      this.startFlushInterval();
    }
  }

  // ===========================================================================
  // Logging Methods
  // ===========================================================================

  /**
   * Log an audit entry
   */
  async log(entry: AuditLogEntry): Promise<void> {
    if (!this.config.enabled) return;
    if (!this.shouldLog(entry.level || 'info')) return;

    const sanitized = this.sanitizeEntry(entry);
    this.queue.push(sanitized);

    // Flush immediately if queue is large
    if (this.queue.length >= this.BATCH_SIZE) {
      await this.flush();
    }
  }

  /**
   * Log an info-level entry
   */
  async info(entry: Omit<AuditLogEntry, 'level'>): Promise<void> {
    return this.log({ ...entry, level: 'info' });
  }

  /**
   * Log a warning-level entry
   */
  async warning(entry: Omit<AuditLogEntry, 'level'>): Promise<void> {
    return this.log({ ...entry, level: 'warning' });
  }

  /**
   * Log an error-level entry
   */
  async error(entry: Omit<AuditLogEntry, 'level'>): Promise<void> {
    return this.log({ ...entry, level: 'error' });
  }

  /**
   * Log a security-level entry
   */
  async security(entry: Omit<AuditLogEntry, 'level'>): Promise<void> {
    return this.log({ ...entry, level: 'security' });
  }

  // ===========================================================================
  // Query Methods
  // ===========================================================================

  /**
   * Query audit logs with filtering and pagination
   */
  async query(options: AuditQueryOptions = {}): Promise<AuditQueryResult> {
    const {
      startDate,
      endDate,
      userId,
      level,
      levels,
      action,
      actions,
      category,
      categories,
      search,
      page = 1,
      limit = 50,
      sortBy = 'timestamp',
      sortOrder = 'desc',
    } = options;

    const where: Record<string, unknown> = {};

    // Date range filter
    if (startDate || endDate) {
      where.timestamp = {};
      if (startDate) (where.timestamp as Record<string, Date>).gte = startDate;
      if (endDate) (where.timestamp as Record<string, Date>).lte = endDate;
    }

    // User filter
    if (userId) where.userId = userId;

    // Level filter (single or multiple)
    if (level) where.level = level;
    if (levels?.length) where.level = { in: levels };

    // Action filter (single or multiple)
    if (action) where.action = { contains: action };
    if (actions?.length) where.action = { in: actions };

    // Category filter (single or multiple)
    if (category) where.category = category;
    if (categories?.length) where.category = { in: categories };

    // Search filter (searches across action, path, userEmail)
    if (search) {
      where.OR = [
        { action: { contains: search, mode: 'insensitive' } },
        { path: { contains: search, mode: 'insensitive' } },
        { userEmail: { contains: search, mode: 'insensitive' } },
      ];
    }

    const [logs, total] = await Promise.all([
      this.prisma.auditLog.findMany({
        where,
        orderBy: { [sortBy]: sortOrder },
        skip: (page - 1) * limit,
        take: limit,
      }),
      this.prisma.auditLog.count({ where }),
    ]);

    // Convert metadata from JSON string to object if needed
    const formattedLogs: AuditLogRecord[] = logs.map((log) => ({
      ...log,
      metadata: typeof log.metadata === 'string'
        ? JSON.parse(log.metadata)
        : log.metadata as Record<string, unknown> | undefined,
    }));

    return {
      logs: formattedLogs,
      total,
      page,
      limit,
      totalPages: Math.ceil(total / limit),
    };
  }

  /**
   * Get a single audit log by ID
   */
  async getById(id: string): Promise<AuditLogRecord | null> {
    const log = await this.prisma.auditLog.findUnique({ where: { id } });

    if (!log) return null;

    return {
      ...log,
      metadata: typeof log.metadata === 'string'
        ? JSON.parse(log.metadata)
        : log.metadata as Record<string, unknown> | undefined,
    };
  }

  /**
   * Get audit log statistics
   */
  async getStats(days: number = 30): Promise<AuditStats> {
    const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000);

    const [total, byLevelRaw, byCategoryRaw] = await Promise.all([
      this.prisma.auditLog.count({ where: { timestamp: { gte: startDate } } }),
      this.prisma.auditLog.groupBy({
        by: ['level'],
        _count: true,
        where: { timestamp: { gte: startDate } },
      }),
      this.prisma.auditLog.groupBy({
        by: ['category'],
        _count: true,
        where: { timestamp: { gte: startDate } },
      }),
    ]);

    // Build byLevel object
    const byLevel: Record<AuditLevel, number> = {
      info: 0,
      warning: 0,
      error: 0,
      security: 0,
    };
    for (const item of byLevelRaw) {
      if (item.level && item.level in byLevel) {
        byLevel[item.level as AuditLevel] = item._count;
      }
    }

    // Build byCategory object
    const byCategory: Record<string, number> = {};
    for (const item of byCategoryRaw) {
      if (item.category) {
        byCategory[item.category] = item._count;
      }
    }

    // Get daily activity counts for the last N days using raw query
    const recentActivity: Array<{ date: string; count: number }> = [];
    try {
      const dailyCounts = await this.prisma.$queryRaw<
        Array<{ date: string; count: bigint }>
      >`
        SELECT DATE(timestamp) as date, COUNT(*) as count
        FROM "AuditLog"
        WHERE timestamp >= ${startDate}
        GROUP BY DATE(timestamp)
        ORDER BY date DESC
        LIMIT 30
      `;

      for (const row of dailyCounts) {
        recentActivity.push({
          date: row.date,
          count: Number(row.count),
        });
      }
    } catch (err) {
      // Raw query may not work with all database providers, fallback to empty array
      console.warn('[AuditService] Failed to get daily counts:', err);
    }

    return {
      total,
      byLevel,
      byCategory,
      recentActivity,
    };
  }

  // ===========================================================================
  // Maintenance Methods
  // ===========================================================================

  /**
   * Delete logs older than retention period
   */
  async cleanup(): Promise<number> {
    const cutoffDate = new Date(
      Date.now() - this.config.retentionDays * 24 * 60 * 60 * 1000
    );

    const result = await this.prisma.auditLog.deleteMany({
      where: { timestamp: { lt: cutoffDate } },
    });

    return result.count;
  }

  /**
   * Export logs to JSON format
   */
  async exportToJson(options: AuditQueryOptions): Promise<string> {
    const result = await this.query({ ...options, limit: 10000 });
    return JSON.stringify(result.logs, null, 2);
  }

  /**
   * Export logs to CSV format
   */
  async exportToCsv(options: AuditQueryOptions): Promise<string> {
    const result = await this.query({ ...options, limit: 10000 });

    const headers = [
      'id',
      'timestamp',
      'level',
      'action',
      'category',
      'userId',
      'userEmail',
      'method',
      'path',
      'statusCode',
      'ipAddress',
      'duration',
      'error',
    ];

    const rows = result.logs.map((log) =>
      headers.map((h) => {
        const value = log[h as keyof AuditLogRecord];
        if (value === null || value === undefined) return '';
        if (value instanceof Date) return value.toISOString();
        if (typeof value === 'object') return JSON.stringify(value);
        return String(value).replace(/"/g, '""');
      })
    );

    const csvRows = [
      headers.join(','),
      ...rows.map((row) => row.map((cell) => `"${cell}"`).join(',')),
    ];

    return csvRows.join('\n');
  }

  // ===========================================================================
  // Path Checking
  // ===========================================================================

  /**
   * Check if a path should be excluded from logging
   */
  shouldExcludePath(path: string): boolean {
    return this.config.excludePaths.some((pattern) => pattern.test(path));
  }

  // ===========================================================================
  // Private Methods
  // ===========================================================================

  private shouldLog(level: AuditLevel): boolean {
    return this.config.levels.includes(level);
  }

  private sanitizeEntry(entry: AuditLogEntry): AuditLogEntry {
    const sanitized = { ...entry };

    // Mask sensitive fields in metadata
    if (sanitized.metadata) {
      sanitized.metadata = this.maskSensitiveData(sanitized.metadata);
    }

    return sanitized;
  }

  private maskSensitiveData(data: Record<string, unknown>): Record<string, unknown> {
    const masked: Record<string, unknown> = {};

    for (const [key, value] of Object.entries(data)) {
      if (this.config.maskFields.some((field) =>
        key.toLowerCase().includes(field.toLowerCase())
      )) {
        masked[key] = '[REDACTED]';
      } else if (typeof value === 'object' && value !== null) {
        masked[key] = this.maskSensitiveData(value as Record<string, unknown>);
      } else {
        masked[key] = value;
      }
    }

    return masked;
  }

  private startFlushInterval(): void {
    this.flushInterval = setInterval(() => {
      this.flush().catch((err) => {
        console.error('[AuditService] Flush error:', err);
      });
    }, this.FLUSH_INTERVAL_MS);
  }

  private async flush(): Promise<void> {
    if (this.queue.length === 0) return;

    const entries = this.queue.splice(0, this.BATCH_SIZE);

    try {
      await this.prisma.auditLog.createMany({
        data: entries.map((entry) => ({
          level: entry.level || 'info',
          action: entry.action,
          category: entry.category || 'api',
          userId: entry.userId,
          userEmail: entry.userEmail,
          targetId: entry.targetId,
          targetType: entry.targetType,
          method: entry.method,
          path: entry.path,
          statusCode: entry.statusCode,
          ipAddress: entry.ipAddress,
          userAgent: entry.userAgent,
          duration: entry.duration,
          metadata: entry.metadata ? entry.metadata : undefined,
          error: entry.error,
          timestamp: new Date(),
        })),
      });
    } catch (err) {
      // Re-add entries to queue on failure
      this.queue.unshift(...entries);
      throw err;
    }
  }

  private parseLevels(levelsStr?: string): AuditLevel[] {
    if (!levelsStr) {
      return ['info', 'warning', 'error', 'security'];
    }
    return levelsStr.split(',').map((l) => l.trim() as AuditLevel);
  }

  private parsePaths(pathsStr?: string): RegExp[] {
    if (!pathsStr) {
      return [/^\/health$/, /^\/metrics$/];
    }
    return pathsStr.split(',').map((p) => new RegExp(p.trim()));
  }

  private parseFields(fieldsStr?: string): string[] {
    if (!fieldsStr) {
      return ['password', 'token', 'secret', 'apiKey', 'creditCard', 'ssn'];
    }
    return fieldsStr.split(',').map((f) => f.trim());
  }

  /**
   * Stop the flush interval (for cleanup)
   */
  destroy(): void {
    if (this.flushInterval) {
      clearInterval(this.flushInterval);
      this.flushInterval = null;
    }
    // Flush remaining entries
    this.flush().catch(console.error);
  }
}

// =============================================================================
// Factory
// =============================================================================

let auditServiceInstance: AuditService | null = null;
let sharedPrismaClient: PrismaClient | null = null;

/**
 * Initialize the audit service with a shared Prisma client
 * Call this once during application startup
 */
export function initAuditService(prisma: PrismaClient, config?: Partial<AuditConfig>): AuditService {
  sharedPrismaClient = prisma;
  auditServiceInstance = new AuditService(config, prisma);
  return auditServiceInstance;
}

/**
 * Get or create the audit service singleton
 */
export function getAuditService(): AuditService {
  if (!auditServiceInstance) {
    auditServiceInstance = new AuditService(undefined, sharedPrismaClient || undefined);
  }
  return auditServiceInstance;
}

/**
 * Create a custom audit service instance
 */
export function createAuditService(config: Partial<AuditConfig>, prisma?: PrismaClient): AuditService {
  return new AuditService(config, prisma || sharedPrismaClient || undefined);
}

export default AuditService;

File: modules/audit-log/mobile/lib/core/services/audit_service.dart
import 'package:dartz/dartz.dart';
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

// Import from your app's core package
// Adjust these paths based on your project structure
// Example: import 'package:your_app/core/errors/failures.dart';
import '../errors/failures.dart';
import '../network/api_client.dart';
import '../../data/repositories/base_repository.dart';

// =============================================================================
// Models
// =============================================================================

/// Audit log level enum
enum AuditLevel {
  info,
  warning,
  error,
  security;

  String get value => name;

  static AuditLevel fromString(String value) {
    return AuditLevel.values.firstWhere(
      (e) => e.name == value.toLowerCase(),
      orElse: () => AuditLevel.info,
    );
  }
}

/// Audit log category enum
enum AuditCategory {
  auth,
  user,
  admin,
  payment,
  system,
  api,
  security;

  String get value => name;

  static AuditCategory fromString(String value) {
    return AuditCategory.values.firstWhere(
      (e) => e.name == value.toLowerCase(),
      orElse: () => AuditCategory.api,
    );
  }
}

/// Audit log entry model
class AuditLog {
  final String id;
  final DateTime timestamp;
  final AuditLevel level;
  final String action;
  final String? category;
  final String? userId;
  final String? userEmail;
  final String? targetId;
  final String? targetType;
  final String? method;
  final String? path;
  final int? statusCode;
  final String? ipAddress;
  final String? userAgent;
  final int? duration;
  final Map<String, dynamic>? metadata;
  final String? error;

  const AuditLog({
    required this.id,
    required this.timestamp,
    required this.level,
    required this.action,
    this.category,
    this.userId,
    this.userEmail,
    this.targetId,
    this.targetType,
    this.method,
    this.path,
    this.statusCode,
    this.ipAddress,
    this.userAgent,
    this.duration,
    this.metadata,
    this.error,
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) {
    return AuditLog(
      id: json['id'] as String,
      timestamp: DateTime.parse(json['timestamp'] as String),
      level: AuditLevel.fromString(json['level'] as String? ?? 'info'),
      action: json['action'] as String,
      category: json['category'] as String?,
      userId: json['userId'] as String?,
      userEmail: json['userEmail'] as String?,
      targetId: json['targetId'] as String?,
      targetType: json['targetType'] as String?,
      method: json['method'] as String?,
      path: json['path'] as String?,
      statusCode: json['statusCode'] as int?,
      ipAddress: json['ipAddress'] as String?,
      userAgent: json['userAgent'] as String?,
      duration: json['duration'] as int?,
      metadata: json['metadata'] as Map<String, dynamic>?,
      error: json['error'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'timestamp': timestamp.toIso8601String(),
      'level': level.value,
      'action': action,
      'category': category,
      'userId': userId,
      'userEmail': userEmail,
      'targetId': targetId,
      'targetType': targetType,
      'method': method,
      'path': path,
      'statusCode': statusCode,
      'ipAddress': ipAddress,
      'userAgent': userAgent,
      'duration': duration,
      'metadata': metadata,
      'error': error,
    };
  }
}

/// Audit logs query result
class AuditLogsResult {
  final List<AuditLog> logs;
  final int total;
  final int page;
  final int limit;
  final int totalPages;

  const AuditLogsResult({
    required this.logs,
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
  });

  factory AuditLogsResult.fromJson(Map<String, dynamic> json) {
    return AuditLogsResult(
      logs: (json['logs'] as List<dynamic>)
          .map((e) => AuditLog.fromJson(e as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      page: json['page'] as int,
      limit: json['limit'] as int,
      totalPages: json['totalPages'] as int,
    );
  }
}

/// Audit log statistics
class AuditStats {
  final int total;
  final Map<String, int> byLevel;
  final Map<String, int> byCategory;
  final List<DailyActivity> recentActivity;

  const AuditStats({
    required this.total,
    required this.byLevel,
    required this.byCategory,
    required this.recentActivity,
  });

  factory AuditStats.fromJson(Map<String, dynamic> json) {
    return AuditStats(
      total: json['total'] as int,
      byLevel: Map<String, int>.from(json['byLevel'] as Map),
      byCategory: Map<String, int>.from(json['byCategory'] as Map),
      recentActivity: (json['recentActivity'] as List<dynamic>?)
              ?.map((e) => DailyActivity.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

/// Daily activity count
class DailyActivity {
  final String date;
  final int count;

  const DailyActivity({
    required this.date,
    required this.count,
  });

  factory DailyActivity.fromJson(Map<String, dynamic> json) {
    return DailyActivity(
      date: json['date'] as String,
      count: json['count'] as int,
    );
  }
}

/// Query options for fetching audit logs
class AuditQueryOptions {
  final DateTime? startDate;
  final DateTime? endDate;
  final String? userId;
  final AuditLevel? level;
  final List<AuditLevel>? levels;
  final String? action;
  final AuditCategory? category;
  final List<AuditCategory>? categories;
  final String? search;
  final int page;
  final int limit;
  final String sortBy;
  final String sortOrder;

  const AuditQueryOptions({
    this.startDate,
    this.endDate,
    this.userId,
    this.level,
    this.levels,
    this.action,
    this.category,
    this.categories,
    this.search,
    this.page = 1,
    this.limit = 50,
    this.sortBy = 'timestamp',
    this.sortOrder = 'desc',
  });

  Map<String, String> toQueryParams() {
    final params = <String, String>{
      'page': page.toString(),
      'limit': limit.toString(),
      'sortBy': sortBy,
      'sortOrder': sortOrder,
    };

    if (startDate != null) {
      params['startDate'] = startDate!.toIso8601String();
    }
    if (endDate != null) {
      params['endDate'] = endDate!.toIso8601String();
    }
    if (userId != null) {
      params['userId'] = userId!;
    }
    if (level != null) {
      params['level'] = level!.value;
    }
    if (levels != null && levels!.isNotEmpty) {
      params['levels'] = levels!.map((l) => l.value).join(',');
    }
    if (action != null) {
      params['action'] = action!;
    }
    if (category != null) {
      params['category'] = category!.value;
    }
    if (categories != null && categories!.isNotEmpty) {
      params['categories'] = categories!.map((c) => c.value).join(',');
    }
    if (search != null && search!.isNotEmpty) {
      params['search'] = search!;
    }

    return params;
  }

  AuditQueryOptions copyWith({
    DateTime? startDate,
    DateTime? endDate,
    String? userId,
    AuditLevel? level,
    List<AuditLevel>? levels,
    String? action,
    AuditCategory? category,
    List<AuditCategory>? categories,
    String? search,
    int? page,
    int? limit,
    String? sortBy,
    String? sortOrder,
  }) {
    return AuditQueryOptions(
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      userId: userId ?? this.userId,
      level: level ?? this.level,
      levels: levels ?? this.levels,
      action: action ?? this.action,
      category: category ?? this.category,
      categories: categories ?? this.categories,
      search: search ?? this.search,
      page: page ?? this.page,
      limit: limit ?? this.limit,
      sortBy: sortBy ?? this.sortBy,
      sortOrder: sortOrder ?? this.sortOrder,
    );
  }
}

// =============================================================================
// API Constants
// =============================================================================

/// Audit log API endpoints
class AuditApiConstants {
  static const String basePath = '/admin/audit-logs';
  static const String logs = basePath;
  static const String stats = '$basePath/stats';
  static const String export = '$basePath/export';
  static String logById(String id) => '$basePath/$id';
}

// =============================================================================
// Repository
// =============================================================================

/// Audit log repository provider
final auditRepositoryProvider = Provider<AuditRepository>((ref) {
  return AuditRepositoryImpl(dio: ref.watch(dioProvider));
});

/// Audit log repository interface
abstract class AuditRepository {
  /// Get audit logs with optional filters
  Future<Either<Failure, AuditLogsResult>> getLogs(AuditQueryOptions options);

  /// Get a single audit log by ID
  Future<Either<Failure, AuditLog>> getLogById(String id);

  /// Get audit log statistics
  Future<Either<Failure, AuditStats>> getStats({int days = 30});

  /// Export audit logs (returns download URL or file content)
  Future<Either<Failure, String>> exportLogs({
    required String format,
    AuditQueryOptions? options,
  });
}

/// Audit log repository implementation
class AuditRepositoryImpl with BaseRepository implements AuditRepository {
  final Dio _dio;

  AuditRepositoryImpl({required Dio dio}) : _dio = dio;

  @override
  Future<Either<Failure, AuditLogsResult>> getLogs(
    AuditQueryOptions options,
  ) async {
    return safeCall(() async {
      final response = await _dio.get(
        AuditApiConstants.logs,
        queryParameters: options.toQueryParams(),
      );

      final data = response.data as Map<String, dynamic>;
      if (data['success'] != true) {
        throw Exception(data['error'] ?? 'Failed to fetch audit logs');
      }

      return AuditLogsResult.fromJson(data);
    });
  }

  @override
  Future<Either<Failure, AuditLog>> getLogById(String id) async {
    return safeCall(() async {
      final response = await _dio.get(AuditApiConstants.logById(id));

      final data = response.data as Map<String, dynamic>;
      if (data['success'] != true) {
        throw Exception(data['error'] ?? 'Failed to fetch audit log');
      }

      return AuditLog.fromJson(data['log'] as Map<String, dynamic>);
    });
  }

  @override
  Future<Either<Failure, AuditStats>> getStats({int days = 30}) async {
    return safeCall(() async {
      final response = await _dio.get(
        AuditApiConstants.stats,
        queryParameters: {'days': days.toString()},
      );

      final data = response.data as Map<String, dynamic>;
      if (data['success'] != true) {
        throw Exception(data['error'] ?? 'Failed to fetch audit statistics');
      }

      return AuditStats.fromJson(data['stats'] as Map<String, dynamic>);
    });
  }

  @override
  Future<Either<Failure, String>> exportLogs({
    required String format,
    AuditQueryOptions? options,
  }) async {
    return safeCall(() async {
      final queryParams = <String, String>{
        'format': format,
        ...(options?.toQueryParams() ?? {}),
      };

      final response = await _dio.get(
        AuditApiConstants.export,
        queryParameters: queryParams,
        options: Options(responseType: ResponseType.plain),
      );

      return response.data as String;
    });
  }
}
